services/user/index.js

```diff
@@ -1,44 +1,131 @@
-const express = require('express')
-const router = express.Router()
+const express = require("express");
+const router = express.Router();
+const pool = require("../../db");
+const { generateToken, verifyToken } = require("../../authentication");
 
 /**
  * Returns the user list.
- * Observe good security practices such as removing the password field 
+ * Observe good security practices such as removing the password field
  * from the response object.
- * 
+ *
  * If :id is given, the user with the given id will be returned.
  */
-router.get('/', (request, response) => {
-    response.status(500).send({
-        error: 'Pending Implementations'
-    })
-})
+router.get("/", verifyToken, async (request, response) => {
+  try {
+    const { rows: users } = await pool.query(
+      "SELECT id, first_name, last_name, username FROM users"
+    );
+    response.status(200).json(users);
+  } catch (error) {
+    response.status(500).send({ error });
+  }
+});
 
 /**
  * Returns the user with the given id.
  */
-router.get('/:id', (request, response) => {
-    response.status(500).send({
-        error: 'Pending Implementations'
-    })
-})
+router.get("/:id", verifyToken, async (request, response) => {
+  try {
+    const { id } = request.params;
+    const {
+      rows: [user],
+    } = await pool.query(
+      "SELECT id, first_name, last_name, username FROM users WHERE id = $1",
+      [id]
+    );
+    if (!user) return response.status(404).send({ error: "user not found" });
+
+    response.status(200).json(user);
+  } catch (error) {
+    response.status(500).send({ error });
+  }
+});
 
 /**
  * Registers a new user.
  */
-router.post('/register', (request, response) => {
-    response.status(500).send({
-        error: 'Pending Implementations'
-    })
-})
+router.post("/register", async (request, response) => {
+  try {
+    const { first_name, last_name, username, password } = request.body;
+    if (!(first_name && last_name && username && password))
+      return response.status(400).send({ message: "all inputs are required" });
+
+    const {
+      rows: [user],
+    } = await pool.query(
+      "INSERT INTO users (first_name, last_name, username, password) VALUES ($1, $2, $3, crypt($4, gen_salt('bf'))) RETURNING first_name, last_name, username, password",
+      [first_name, last_name, username, password]
+    );
+    response.status(200).json(user);
+  } catch (error) {
+    response.status(500).send({ error });
```

Comment:
Error response data is a little scary in a way that it reveals sensitive info like table and field names. This might hold true for your other handled errors as well.

A common pitfall, but something to seriously consider when implementing production systems.

You may want to do some sort of dictionary for your errors and respond accordingly such as:
`{ success: false, message: "Username already exists" }`

![image](https://user-images.githubusercontent.com/8315544/150803380-1bbb2b86-4f5b-4860-aba0-b26963ec7705.png)